src/plugins/discover/public/context_awareness/profile_service.ts

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+/* eslint-disable max-classes-per-file */
+
+import type { ComposableProfile, PartialProfile } from './composable_profile';
+import type { Profile } from './types';
+
+export type ResolveProfileResult<TContext> =
+  | { isMatch: true; context: TContext }
+  | { isMatch: false };
+
+export type ProfileProviderMode = 'sync' | 'async';
+
+export interface ProfileProvider<
+  TProfile extends PartialProfile,
+  TParams,
+  TContext,
+  TMode extends ProfileProviderMode
+> {
+  order: number;
+  profileId: string;
+  profile: ComposableProfile<TProfile>;
+  resolve: (
+    params: TParams
+  ) => TMode extends 'sync'
+    ? ResolveProfileResult<TContext>
+    : ResolveProfileResult<TContext> | Promise<ResolveProfileResult<TContext>>;
+}
+
+export type ContextWithProfileId<TContext> = TContext & { profileId: string };
+
+const EMPTY_PROFILE = {};
+
+abstract class BaseProfileService<
+  TProfile extends PartialProfile,
+  TParams,
+  TContext,
+  TMode extends ProfileProviderMode
+> {
+  protected readonly providers: Array<ProfileProvider<TProfile, TParams, TContext, TMode>> = [];
+
+  protected constructor(public readonly defaultContext: ContextWithProfileId<TContext>) {}
+
+  public registerProvider(provider: ProfileProvider<TProfile, TParams, TContext, TMode>) {
+    this.providers.push(provider);
+    this.providers.sort((a, b) => a.order - b.order);
+  }
+
+  public getProfile(context: ContextWithProfileId<TContext>): ComposableProfile<Profile> {
+    const provider = this.providers.find((current) => current.profileId === context.profileId);
+    return provider?.profile ?? EMPTY_PROFILE;
+  }
```

Comment:
**doubt**: I might be missing something here, but could you explain a use case of where/why we need an order property for the registered profiles? The only case I can think of is in case we have the exact same resolved value in a specific part of the page (e.g. top nav items), but in this case shouldn't we enforce a constrained set of `profileId` values to register a profile?

Taken from the example profile, where are we potentially going to register a profile with id `logs-data-source-profile` more than once?

Reply:
The `order` prop determines the order for providers to run in during context resolution since the first match is the one that gets used. So for example if we have a `logsDataSourceProfileProvider` and a `metricsDataSourceProfileProvider`, the `order` will determine which one tries to resolve first.

But this was done originally under the assumption we would expose the registration function externally, which may not be necessary if we just do all registrations internally in Discover. In that case, I think it would make sense to just run the providers in whichever order they were registered in code. WDYT?

Reply:
> The order prop determines the order for providers to run in during context resolution since the first match is the one that gets used. So for example if we have a logsDataSourceProfileProvider and a metricsDataSourceProfileProvider, the order will determine which one tries to resolve first.

This is the part where I have doubts, for the `order` value to have a role in the resolution step, we are assuming that there might be multiple registration of profiles with id `logsDataSourceProfileProvider` or anything else, what is the use case where we need to register multiple profiles for logs data source? Or is it just an example id and I am misleading the change that we can have profiles with same ids for other features?

Reply:
Bad explanation on my part, let me try again 😄

The `order` property was used to determine the resolution order for different providers of the same type (e.g. `RootProfileProvider`, `DataSourceProfileProvider`, `DocumentProfileProvider`), unrelated to `profileId`, which as you would imagine should be unique. For example:
```ts
// Set order in providers
const logsDataSourceProfileProvider: DataSourceProfileProvider = { order: 0, profileId: 'logs-data-source-profile', ... };
const metricsDataSourceProfileProvider: DataSourceProfileProvider = { order: 100, profileId: 'metrics-data-source-profile', ... };

// `resolve` uses `order` to decide provider resolution order when looking for a matching profile
const context = await dataSourceProfileProvider.resolve({ ... });
```

The `profileId` is used to map a resolved context to its corresponding profile so we can look it up later, e.g. to get the `DocumentProfile` associated with a data table record:
```ts
// `getProfile` uses the `profileId` property of `record.context` to look up its profile
const documentProfile = documentProfileService.getProfile(record.context);
```

That said, the conversations I've had about it made me realize that `order` is confusing and unnecessary assuming we intend to register providers from within the Discover codebase, in which case we can just rely on registration order, so I removed it here: https://github.com/elastic/kibana/pull/183797/commits/e953b4126bf234d34fa19dde951f16a1a1042e9a.